File: src/bibliography/citeproc.py
from json import dumps, loads
from os.path import join, dirname
from subprocess import call, Popen, PIPE

class Citeproc:

    def call(self, stdin):
        script = join(dirname(__file__), 'citeproc.js')
        node = Popen(['node', script], stdout=PIPE, stdin=PIPE)
        node.stdin.write(stdin.encode())
        stdout, errors = node.communicate()
        return stdout.decode('utf-8')

    def render(self, references):
        json_references = dumps(references)
        output = self.call(json_references)
        metadata, bibliography = loads(output)
        references = self.flatten(metadata['entry_ids'])
        return list(zip(references, bibliography))

    @staticmethod
    def flatten(iterable):
        return [item[0] for item in iterable]
        

File: src/bibliography/citeproc.js
const fs = require('fs');
const CSL = require('../../lib/citeproc-js/citeproc.js').CSL;

let buffer = '';
process.stdin.on('data', chunk => buffer += chunk);

const generateBibliography = references => {
  const citeprocSys = {
    retrieveItem: (id) =>
      references.find(reference => reference.id == id),
    retrieveStyle: (style) =>
        //fs.readFileSync(__dirname + '/styles/apa.csl', 'utf-8'),
        fs.readFileSync(__dirname + '/styles/harvard-cite-them-right.csl', 'utf-8'),
    retrieveLocale: (lang) => 
      fs.readFileSync(__dirname + '/locales/locales-nb-NO.xml', 'utf-8')
  };

  const engine = new CSL.Engine(citeprocSys, citeprocSys.retrieveStyle());
  engine.updateItems(references.map(reference => reference.id));
  const bibliography = engine.makeBibliography();
  return bibliography; 
}

process.stdin.on('end', () => {
  const references = JSON.parse(buffer);
  const bibliography = generateBibliography(references);
  const output = JSON.stringify(bibliography);
  process.stdout.write(output);
});
